[VIKI] Fixed login and Implemented new method for stream request using android api

zackmark29 opened the pull request:
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [ ] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [ ] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [ ] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [ ] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [x] Improvement

---

### To do
- [x] Fix channel request

### Description of your *pull request* and other information

Explanation of your *pull request* in arbitrary form goes here. Please make sure the description explains the purpose and effect of your *pull request* and is worded well enough to be understood. Provide as much context and examples as possible.

pukkandan commented on yt_dlp/extractor/viki.py:
```diff
@@ -295,15 +309,35 @@ def _real_extract(self, url):
+        stream_url = 'playback_streams/%s.json?drms=dt1,dt2&device_id=%s&token=%s' % (video_id, self._DEVICE_ID, self._token)
+        resp = self._call_api(stream_url, video_id, 'Downloading video streams JSON')
 
+        stream_id = resp['main'][0]['properties']['track']['stream_id']
```
Use `try_get`. If subs cant be found, the rest of the extraction should still run error-free

zackmark29 replied:
it's always present it the response.

rautamiekka replied:
Don't jinx it, it's been seen there's no such thing as absolute certainty in websites.

pukkandan commented on yt_dlp/extractor/viki.py:
```diff
@@ -295,15 +309,35 @@ def _real_extract(self, url):
+        self._sort_formats(formats)
+        
+        
         result = {
```
You can return it directly here:
```py
return {
    'id': video_id,
    'formats': formats,
    'title': title,
    ...
}
```

zackmark29 commented on yt_dlp/extractor/viki.py:
```diff
@@ -295,15 +309,35 @@ def _real_extract(self, url):
 
+        mpd_url = self.search_another_mpd_url(resp['main'][0]['url'], video_id)
+
+        '''
+        NOTE: audio formats should be sorted by height which can get from audio dash url.
+        '''
```
It's just to sort the high audio from the audio url.
because of this:
They have same bitrate. When you sort it, it will choose the id "8" instead of 9.

`<Representation id="8" mimeType="audio/mp4" codecs="mp4a.40.2" audioSamplingRate="44100" startWithSAP="1" bandwidth="194473">
        <AudioChannelConfiguration schemeIdUri="urn:mpeg:dash:23003:3:audio_channel_configuration:2011" value="2"/>
        <BaseURL>1183352v_dash_high_1080p_2a98c0_2106140900_track2_dashinit.mp4</BaseURL>
        <SegmentBase indexRangeExact="true" indexRange="853-10364">
          <Initialization range="0-852"/>
        </SegmentBase>
      </Representation>
      <Representation id="9" mimeType="audio/mp4" codecs="mp4a.40.2" audioSamplingRate="44100" startWithSAP="1" bandwidth="194473">
        <AudioChannelConfiguration schemeIdUri="urn:mpeg:dash:23003:3:audio_channel_configuration:2011" value="2"/>
        <BaseURL>1183352v_dash_high_720p_00c68c_2106140800_track2_dashinit.mp4</BaseURL>
        <SegmentBase indexRangeExact="true" indexRange="853-10364">
          <Initialization range="0-852"/>`

zackmark29 commented on yt_dlp/extractor/viki.py:
```diff
@@ -295,15 +309,35 @@ def _real_extract(self, url):
+                'url': subtitle_format % (video_id, subtitle_lang, subtitle_format, self._APP, stream_id)
             } for subtitles_format in ('srt', 'vtt')]
 
+        mpd_url = self.search_another_mpd_url(resp['main'][0]['url'], video_id)
```
Because it's need to check for another mpd url which is have max 1080 format.
The first mpd only sd formats.

zackmark29 commented on yt_dlp/extractor/viki.py:
```diff
@@ -46,47 +48,49 @@ class VikiBaseIE(InfoExtractor):
+            
+        query = self._API_QUERY_TEMPLATE % (self._API_VER_5, path, self._APP)
         if self._token:
             query += '&token=%s' % self._token
         sig = hmac.new(
-            self._APP_SECRET.encode('ascii'),
-            query.encode('ascii'),
+            self._APP_SECRET.encode('ascii'), f'{query}&t={timestamp}'.encode('ascii'),
```
&t= is only required for signing and not part of the api url. That's why i removed it from template

zackmark29 commented on yt_dlp/extractor/viki.py:
```diff
@@ -295,15 +309,35 @@ def _real_extract(self, url):
+        NOTE: audio formats should be sorted by height which can get from audio dash url.
+        '''
+        formats = []
+        formats.append({
+            'url': mpd_url,
+            'format_id': '%s-%s' % ('dash', 'https'),
+        })
+        formats.extend(self._extract_mpd_formats(mpd_url, video_id))
```
what do you mean? this 'url': mpd_url?

zackmark29 commented:
I'm wondering about this, why need to literate also the clips and movies? 
can I separate this for movies and series channel?
```
entries = []
for video_type in ('episodes', 'clips', 'movies'):
    for page_num in itertools.count(1):
        page = self._call_api(
            'containers/%s/%s.json?per_page=%d&sort=number&direction=asc&with_paging=true&page=%d'
            % (channel_id, video_type, self._PER_PAGE, page_num), channel_id,
            'Downloading %s JSON page #%d' % (video_type, page_num))
        for video in page['response']:
            video_id = video['id']
            entries.append(self.url_result(
                'https://www.viki.com/videos/%s' % video_id, 'Viki'))
        if not page['pagination']['next']:
                    break```

zackmark29 commented:
fixed for downloading channel and updated metadata api
```python
[viki:channel] Logging in
[debug] [viki:channel] Extracting URL: https://www.viki.com/tv/37776c-doom-at-your-service
[viki:channel] 37776c: Downloading channel JSON
[viki:channel] 37776c: Downloading episodes JSON page #1
[download] Downloading playlist: Doom at Your Service
[viki:channel] playlist Doom at Your Service: Collected 16 videos; downloading 16 of them
[download] Downloading video 1 of 16
[viki] Logging in
[debug] [viki] Extracting URL: https://www.viki.com/videos/1180926v-doom-at-your-service-episode-16
[viki] 1180926v: Downloading video JSON
[viki] 1180926v: Downloading video streams JSON
[viki] 1180926v: Checking for another mpd url
[viki] 1180926v: Downloading MPD manifest
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, vcodec:vp9.2(10), acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[info] 1180926v: Downloading 1 format(s): 0
[debug] Invoking downloader on "https://m-content6-viki.s.llnwi.net/1180926v/dash/1180926v_dash_high_1080p_a678e7_2106291758_track1_dashinit.mp4"
[download] Destination: Doom at Your Service - Episode 16 [1180926v].mp4
[download]   0.0% of 905.91MiB at 68.96KiB/s ETA 03:44:10
```

zackmark29 commented:
@pukkandan kindly check again. I did some changes. It should work all now
Only some clean up are needed

pukkandan commented:
#475
